src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java

```diff
@@ -54,26 +62,81 @@ public NtlmAuthenticator create() {
         }
     }
 
-    public long authenticate(Connection connection, String username, char[] password) throws TransportException {
+    public long authenticate(Connection connection, AuthenticationContext context) throws TransportException {
         try {
-            logger.info("Authenticating {} on {} using NTLM", username, connection.getRemoteHostname());
-            SMB2SessionSetup smb2SessionSetup = new SMB2SessionSetup(connection.getNegotiatedDialect());
+            logger.info("Authenticating {} on {} using NTLM", context.getUsername(), connection.getRemoteHostname());
+            EnumSet<SMB2SessionSetup.SMB2SecurityMode> signingEnabled = EnumSet.of
+                    (SMB2SessionSetup.SMB2SecurityMode.SMB2_NEGOTIATE_SIGNING_ENABLED);
+
+            SMB2SessionSetup smb2SessionSetup = new SMB2SessionSetup(connection.getNegotiatedDialect(), signingEnabled);
             NtlmNegotiate ntlmNegotiate = new NtlmNegotiate();
             byte[] asn1 = negTokenInit(ntlmNegotiate);
             smb2SessionSetup.setSecurityBuffer(asn1);
             connection.send(smb2SessionSetup);
             SMB2SessionSetup receive = (SMB2SessionSetup) connection.receive();
+            long sessionId = receive.getHeader().getSessionId();
             if (receive.getHeader().getStatus() == SMB2StatusCode.STATUS_MORE_PROCESSING_REQUIRED) {
-                logger.debug("More processing required for authentication of {}", username);
+                logger.debug("More processing required for authentication of {}", context.getUsername());
                 byte[] securityBuffer = receive.getSecurityBuffer();
                 logger.info("Received token: {}", ByteArrayUtils.printHex(securityBuffer));
 
                 NegTokenTarg negTokenTarg = new NegTokenTarg().read(securityBuffer);
                 BigInteger negotiationResult = negTokenTarg.getNegotiationResult();
-                NtlmChallenge read = (NtlmChallenge) new NtlmChallenge().read(new Buffer.PlainBuffer(negTokenTarg.getResponseToken(), Endian.LE));
-                logger.debug("Received NTLM challenge from: {}", read.getTargetName());
+                NtlmChallenge challenge = (NtlmChallenge) new NtlmChallenge().read(new Buffer.PlainBuffer(negTokenTarg.getResponseToken(), Endian.LE));
+                logger.debug("Received NTLM challenge from: {}", challenge.getTargetName());
+
+                byte[] serverChallenge = challenge.getServerChallenge();
+
+                byte[] responseKeyNT = NtlmFunctions.NTOWFv2(
+                        String.valueOf(context.getPassword()),
+                        context.getUsername(),
+                        context.getDomain());
+
+                byte[] challengeFromClient = new byte[8];
```

Comment:
Inline this in the `getNTLMv2ClientChallenge`. The `challengeFromClient` is only used there. This also removed the need for the public `RANDOM` constant, which can then be made private.